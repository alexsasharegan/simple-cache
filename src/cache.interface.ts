import { Option } from "safe-types"

export interface BaseCache<K, V> {
	/**
	 * Read a value from the cache by key as an `Option<V>`.
	 */
	get(key: K): Option<V>
	/**
	 * Read a value from the cache by key. May not return a value if cache miss.
	 */
	read(key: K): V | undefined
	/**
	 * Write a value to the cache at a given key.
	 */
	write(key: K, value: V): void
	/**
	 * Delete a keyed value from the cache.
	 */
	remove(key: K): void
	/**
	 * Wipe the cache clean.
	 */
	clear(): void
	/**
	 * Wipe the cache clean. (Alias for clear)
	 */
	invalidate(): void
	/**
	 * Returns the current size of the cache.
	 */
	size(): number
	/**
	 * Returns an array of all the keys in the cache.
	 */
	keys(): K[]
	/**
	 * Returns an array of all the values in the cache.
	 */
	values(): V[]
	/**
	 * Returns an array of [key, value] array tuples
	 */
	entries(): Array<[K, V]>
}

/**
 * Cache is an interface for caching key/value pairs of types `K` and `V`.
 * It's primary functionality includes:
 * - `Cache.write`: write a value at a key
 * - `Cache.read`: read a value from a key
 * - `Cache.get`: read a value as an `Option<V>` (maybe type)
 * - `Cache.remove`: remove a value by key
 * - `Cache.clear`: empty the cache
 */
export interface Cache<K, V> extends BaseCache<K, V> {
	/**
	 * Returns a string representation of the cache with its label, size, and
	 * capacity.
	 */
	toString(): string
	/**
	 * Serializable for JSON.stringify.
	 */
	toJSON(): Array<[K, V]>
}

export interface TypeLabel {
	key: string
	value: string
}
