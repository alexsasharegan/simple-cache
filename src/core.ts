import { BaseCache } from "./cache.interface"
import { Option } from "safe-types"

/**
 * `CacheItem` is a simple wrapper for a value type that allows cache capacity
 * to be managed by persisting items that are accessed most frequently, and
 * removing items that are accessed least frequently.
 */
export interface CacheItem<K, V> {
	hits: number
	key: K
	value: V
}

/**
 * A `Rebalancer` is a method that operates on a cache to ensure it honors
 * its capacity.
 */
type Rebalancer<K> = (newKey: K) => void

export interface CacheInternals<K, V> {
	get_store(): Map<K, CacheItem<K, V>>
}

/**
 * An `ItemFactory` is a function that returns an object implementing the
 * CacheItem structure. Implementers of this function may choose to extend the
 * CacheItem interface to support special cache behaviors.
 */
export type ItemFactory<K, V> = (key: K, value: V) => CacheItem<K, V>

/**
 * Options for extending the base behavior of the core cache implementation.
 */
export interface BaseCacheOptions<K, V> {
	itemFactory?: ItemFactory<K, V>
}

export function create_rebalancer<K, V>(
	c: Map<K, CacheItem<K, V>>,
	capacity: number
): Rebalancer<K> {
	return function rebalance_cache(newKey: K) {
		// Sort items in ascending order of hits.
		let items = Array.from(c.values()).sort((a, b) => a.hits - b.hits)

		while (c.size > capacity) {
			// Pull items off the least accessed side of the array.
			// Use `!` to assert our value is not void.
			// Cache overflow tests verify we can trust this.
			let { key } = items.shift()!
			// Guard against removing the item just written to the cache
			if (key == newKey) {
				continue
			}

			c.delete(key)
		}
	}
}

function baseItemFactory<K, V>(key: K, value: V): CacheItem<K, V> {
	return {
		key,
		value,
		hits: 0,
	}
}

/**
 * CoreCache implements a base set of the Cache interface's methods, while
 * providing a few extra methods to make composition more flexible to other
 * public Cache implementations.
 */
export function CoreCache<K, V>(
	capacity: number,
	opts: BaseCacheOptions<K, V> = {}
): BaseCache<K, V> & CacheInternals<K, V> {
	if (capacity < 1 || !Number.isInteger(capacity)) {
		throw new RangeError(
			`Cache capacity requires an integer value greater than or equal to 1`
		)
	}

	let c: Map<K, CacheItem<K, V>> = new Map()
	let rebalance = create_rebalancer(c, capacity)
	let { itemFactory = baseItemFactory } = opts

	const cache: BaseCache<K, V> & CacheInternals<K, V> = {
		get_store() {
			return c
		},

		get(k) {
			return Option.of(cache.read(k))
		},

		read(k) {
			let item = c.get(k)
			if (!item) {
				return undefined
			}

			item.hits += 1

			return item.value
		},

		write(k, v) {
			let item = itemFactory(k, v)

			let existing = c.get(k)
			if (existing) {
				// Writes that are updates should increment the hit count
				// rather than initialize it.
				item.hits += existing.hits
			}

			c.set(k, item)

			if (c.size > capacity) {
				rebalance(k)
			}
		},

		remove(k) {
			c.delete(k)
		},

		invalidate() {
			cache.clear()
		},

		clear() {
			c.clear()
		},

		size() {
			return c.size
		},

		keys() {
			return Array.from(c.keys())
		},

		values() {
			return Array.from(c.values(), x => x.value)
		},

		entries() {
			return Array.from(c.entries(), ([k, v]) => {
				// Strip out the internals
				// (TS cast is to preserve the correct tuple shape)
				return [k, v.value] as [K, V]
			})
		},
	}

	return cache
}
