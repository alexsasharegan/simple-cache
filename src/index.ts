export { Cache } from "./cache.interface"
export { SimpleCache } from "./simple-cache"
export { EphemeralCache } from "./ephemeral-cache"
