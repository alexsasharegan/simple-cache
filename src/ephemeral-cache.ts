import { Option } from "safe-types"
import { Cache, TypeLabel } from "./cache.interface"
import { CoreCache, CacheItem } from "./core"

interface EphemeralCacheItem<K, V> extends CacheItem<K, V> {
	/**
	 * Unix milliseconds timestamp representing the time at which a value expires.
	 */
	expiration: number
}

/**
 * EphemeralCache extends the Cache interface to store values temporarily.
 * The implementation does not use timers, so expired values are purged on
 * write/read. To avoid blocking read/write with an `O(n)` purge, the purge is
 * deferred to the next tick using a Promise. This means that the `Cache.size`
 * method is non-deterministic--you can write a value that defers a purge, call
 * size and receive `n`, then try to read `n` values, but because they are
 * expired, you read `n - x(expired)` items. This tradeoff maintains expiration
 * accuracy and prevents read/write call stacks from blocking for a cache
 * expiration purge.
 */
export function EphemeralCache<K, V>(
	capacity: number,
	durationMs: number,
	typeLabel: TypeLabel = { key: "any", value: "any" }
): Cache<K, V> {
	if (durationMs < 1 || !Number.isInteger(durationMs)) {
		throw new RangeError(
			`${
				EphemeralCache.name
			} requires an integer value greater than or equal to 1 for capacity and durationMs`
		)
	}

	const {
		clear,
		entries,
		get_store,
		invalidate,
		keys,
		remove,
		size,
		values,
		write,
	} = CoreCache<K, V>(capacity, { itemFactory })

	function itemFactory<K, V>(key: K, value: V): EphemeralCacheItem<K, V> {
		return {
			key,
			value,
			hits: 0,
			expiration: Date.now() + durationMs,
		}
	}

	let lastPurged = 0
	/**
	 * `defer_purge` compares timestamps on `lastPurged`, and if the `durationMs`
	 * has passed, enqueues a purge on the next tick to prevent callers from
	 * blocking the `O(n)` cost (`n` is the current cache size).
	 */
	function defer_purge() {
		if (Date.now() > lastPurged + durationMs) {
			// Promise.resolve will run on the next tick
			Promise.resolve().then(() => {
				let c = get_store() as Map<K, EphemeralCacheItem<K, V>>
				let now = Date.now()

				for (let { expiration, key } of c.values()) {
					if (now > expiration) {
						remove(key)
					}
				}

				lastPurged = now
			})
		}
	}

	const cache: Cache<K, V> = {
		clear,
		entries,
		invalidate,
		keys,
		remove,
		size,
		values,
		get(k) {
			return Option.of(cache.read(k))
		},

		/**
		 * Implements a custom `read` method so time-to-live expiration is honored.
		 */
		read(k) {
			let c = get_store() as Map<K, EphemeralCacheItem<K, V>>
			defer_purge()

			let item = c.get(k)
			if (!item) {
				return undefined
			}

			if (Date.now() > item.expiration) {
				remove(k)
				return undefined
			}

			item.hits += 1

			return item.value
		},

		write(k, v) {
			write(k, v)
			defer_purge()
		},

		toString() {
			return `${EphemeralCache.name /* makes name refactoring simpler */}<${
				typeLabel.key
			}, ${
				typeLabel.value
			}> { size: ${size()}, capacity: ${capacity}, durationMs: ${durationMs} }`
		},

		toJSON() {
			return entries()
		},
	}

	return cache
}
