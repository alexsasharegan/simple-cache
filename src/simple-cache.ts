import { Cache, TypeLabel } from "./cache.interface"
import { CoreCache } from "./core"

/**
 * SimpleCache follows the Cache interface to store values `V` by key `K`.
 *
 * This cache allows writes up to capacity, beyond which least accessed values
 * will be removed.
 */
export function SimpleCache<K, V>(
	capacity: number,
	typeLabel: TypeLabel = { key: "any", value: "any" }
): Cache<K, V> {
	const {
		clear,
		entries,
		get,
		invalidate,
		keys,
		read,
		remove,
		size,
		values,
		write,
	} = CoreCache<K, V>(capacity)

	const cache: Cache<K, V> = {
		clear,
		entries,
		get,
		invalidate,
		keys,
		read,
		remove,
		size,
		values,
		write,
		toString() {
			return `${SimpleCache.name /* makes name refactoring simpler */}<${
				typeLabel.key
			}, ${typeLabel.value}> { size: ${size()}, capacity: ${capacity} }`
		},

		toJSON() {
			return entries()
		},
	}

	return cache
}
