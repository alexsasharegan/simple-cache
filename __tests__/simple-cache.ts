import { SimpleCache } from "../src/simple-cache";

describe("SimpleCache", async () => {
  it("should toString", async () => {
    let c = SimpleCache<string>(1, "Snapshot");

    expect(String(c)).toMatchSnapshot("SimpleCache.toString");
  });

  it("should be iterable", async () => {
    let c = SimpleCache<number>(10, "number");

    for (let i = 1; i <= 10; i += 1) {
      c.write(i.toString(10), i);
    }

    for (let entry of c.entries()) {
      expect(Array.isArray(entry)).toBe(true);
      let [key, value] = entry;
      expect(typeof key).toBe("string");
      expect(typeof value).toBe("number");
    }
  });

  it("should toJSON", async () => {
    let c = SimpleCache<number>(10, "number");

    for (let i = 1; i <= 10; i += 1) {
      c.write(i.toString(10), i);
    }

    expect(JSON.stringify(c, null, 2)).toMatchSnapshot();
  });

  it("should write and read from cache", async () => {
    let test = {
      key: "test",
      value: "pictalk",
    };
    let c = SimpleCache<string>(8);

    expect(c.read(test.key)).toBe(undefined);

    c.write(test.key, test.value);
    expect(c.read(test.key)).toBe(test.value);
  });

  it("should remove an item by key", async () => {
    let test = {
      key: "test",
      value: "pictalk",
    };
    let c = SimpleCache<string>(8);
    c.write(test.key, test.value);
    c.remove(test.key);
    expect(c.read(test.key)).toBe(undefined);
  });

  it("should return the size of the cache", async () => {
    let c = SimpleCache<number>(10);

    for (let i = 1; i <= 10; i += 1) {
      c.write(i.toString(10), i);
    }

    expect(c.size()).toBe(10);
  });

  it("should return the values of the cache", async () => {
    let c = SimpleCache<number>(10);

    for (let i = 1; i <= 10; i += 1) {
      c.write(i.toString(10), i);
    }

    expect(c.values()).toEqual([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
  });

  it("should return the keys of the cache", async () => {
    let c = SimpleCache<number>(10);

    for (let i = 1; i <= 10; i += 1) {
      c.write(i.toString(10), i);
    }

    expect(c.keys()).toEqual([
      "1",
      "2",
      "3",
      "4",
      "5",
      "6",
      "7",
      "8",
      "9",
      "10",
    ]);
  });

  it("should return the entries of the cache", async () => {
    let c = SimpleCache<number>(10);

    for (let i = 1; i <= 10; i += 1) {
      c.write(i.toString(10), i);
    }

    expect(c.entries()).toEqual([
      ["1", 1],
      ["2", 2],
      ["3", 3],
      ["4", 4],
      ["5", 5],
      ["6", 6],
      ["7", 7],
      ["8", 8],
      ["9", 9],
      ["10", 10],
    ]);
  });

  it("should invalidate cache", async () => {
    let c = SimpleCache<number>(10);

    for (let i = 1; i <= 10; i += 1) {
      c.write(i.toString(10), i);
    }

    c.invalidate();
    expect(c.size()).toBe(0);

    for (let i = 1; i <= 10; i += 1) {
      expect(c.read(i.toString(10))).toBe(undefined);
    }
  });

  it("should not overflow the cache size", async () => {
    let c = SimpleCache<number>(10);

    for (let i = 1; i < 1000; i += 1) {
      c.write(i.toString(10), i);
      // write twice to test the overwrite logic
      c.write(i.toString(10), i);
    }

    expect(c.size()).toBe(10);
    expect(c.values().length).toBe(10);
  });

  it("should update values without increasing the size on same key", async () => {
    let c = SimpleCache<string>(10, "string");
    let k = "same key";
    let fn = (n: number) => `updated value: ${n}`;

    for (let i = 1; i <= 10; i += 1) {
      c.write(k, fn(i));
    }

    expect(c.size()).toBe(1);
    expect(c.read(k)).toBe(fn(10));
  });

  it("should not rebalance and remove the newest key item", async () => {
    let c = SimpleCache<number>(10, "number");

    for (let i = 1; i <= 1000; i += 1) {
      c.write(i.toString(10), i);
      c.read(i.toString(10));
    }

    expect(c.size()).toBe(10);
  });

});
