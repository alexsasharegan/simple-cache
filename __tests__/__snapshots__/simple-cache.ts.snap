// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleCache should toJSON 1`] = `
"{
  \\"1\\": 1,
  \\"2\\": 2,
  \\"3\\": 3,
  \\"4\\": 4,
  \\"5\\": 5,
  \\"6\\": 6,
  \\"7\\": 7,
  \\"8\\": 8,
  \\"9\\": 9,
  \\"10\\": 10
}"
`;

exports[`SimpleCache should toString: SimpleCache.toString 1`] = `"SimpleCache<Snapshot> { size: 0, capacity: 1 }"`;
